### PR TITLE
A6: drop object-literal false positives + remove filename workaround (closes #78)

### LINKED ISSUE #78: [Detection] Filter object-literal false positives in AST scanner (A6, replaces #66)
## Summary
Skip AST matches when the method chain appears as a string key inside an object literal, not as the function of a call. Remove the filename-based workaround that currently masks this bug.

**Replaces #66.**

## Why
`{ "chat.completions.create": { ... } }` in a config file is detected as an API call today. The filename-based workaround in `file-discovery.ts:48-59` (excluding `pricing.ts`, `costs.ts`, etc.) hides this — but also hides real calls in any file that happens to use one of those names.

## Current state
`src/ast/call-visitor.ts` has no parent-node guard. `src/scanner/file-discovery.ts:48-59` blocklists pricing/config filenames.

## What to do
1. In `call-visitor.ts`, walk parent nodes before emitting a match. Skip when ancestor is `object`, `array`, `variable_declarator`, or `assignment_expression`. Emit only when ancestor reaches `call_expression`.
2. Remove all filename-based pricing/config patterns from `DEFAULT_IGNORE_PATTERNS` in `file-discovery.ts`.
3. Remove `api/src/config/pricing.ts` from `.recostignore` unless confirmed data-only.

## Acceptance criteria
- [ ] `pricing.ts` fixture with method-chain string keys produces zero findings
- [ ] Real call in `service.ts` to the same method still detected
- [ ] All filename-based workaround patterns removed
- [ ] Existing fingerprint registry tests pass
- [ ] Benchmark (D1) precision improves measurably

## Reference
Full design: https://github.com/recost-dev/extension/blob/main/docs/accuracy/detection.md#a6-filter-object-literal-false-positives-in-ast-scanner

### PR DESCRIPTION
**Pull request by AndresL230**
## Summary

- Adds fixture `src/test/fixtures/a6/{pricing-config.ts, service.ts}` plus `src/test/a6-object-literal-fps.test.ts` demonstrating that method-chain string keys in an object literal must produce zero detections, while a real `client.chat.completions.create(...)` call in a sibling file still emits exactly one openai endpoint.
- Fixes the FP emit site in `src/scanner/core-scanner.ts`: the multi-line regex expansion fallback was re-emitting an AST-covered call at the FIRST line of its 6-line look-ahead window (`import OpenAI from "openai"` on line 1 was absorbing a real `client.chat.completions.create(...)` from line 5 and emitting phantom call sites at lines 1 and 2). Fix: skip multi-line expansion when any line in the window was already AST-covered.
- Removes the 13 filename-based workarounds in `src/scanner/file-discovery.ts` (`pricing.{ts,js,tsx}`, `costs.{ts,js}`, `rates.{ts,js}`, `api-config.{ts,js}`, `provider-config.{ts,js}`, `api-pricing.{ts,js}`) that were masking real calls in any file matching those names. The openai-compatible regex already requires `\s*\(` after the action chain, so object-literal string keys like `"chat.completions.create":` never matched in the first place — the workarounds were paying suppression cost without doing necessary work.
- Updates `benchmark/baseline.json` with the +1.07pp detection precision improvement.

## A6.0 investigation findings

The issue body claimed the FP was in `call-visitor.ts`. Investigation showed `call-visitor.ts` already only emits matches for real `call_expression` AST nodes — never object-literal keys. The pricing-config fixture itself produced **zero** detections from the AST or the regex pass. The phantom emissions were in `service.ts`, at the `import` and `new OpenAI()` lines, caused by the multi-line regex re-emission in `core-scanner.ts`. Fix landed there.

## D1 measurement

```
Prior (baseline.json on main):
  detectionPrecision        0.2988505747126437
  detectionRecall           0.4262295081967213
  providerAttributionAccuracy 0.7959183673469388
  findingPrecision          0.07142857142857142
  findingRecall             0.3333333333333333

New:
  detectionPrecision        0.30952380952380953
  detectionRecall           0.4262295081967213
  providerAttributionAccuracy 0.7959183673469388
  findingPrecision          0.07142857142857142
  findingRecall             0.3333333333333333
```

Δ (pp) from the runner output:

| Metric                | Δ (pp)  |
| --------------------- | ------- |
| Detection precision   | +1.07   |
| Detection recall      | +0.00   |
| Provider attribution  | +0.00   |
| Finding precision     | +0.00   |
| Finding recall        | +0.00   |

## Test plan

- [ ] CI `benchmark` workflow passes (no metric drops > 1pp from new baseline)
- [ ] CI `test:scanner` passes (4 new `a6-*` tests + 37 existing)
- [ ] Manual: scan a repo with a pricing-style config file containing method-chain string keys — produces zero false positives, and a sibling file with a real `client.chat.completions.create(...)` call emits exactly one endpoint

Closes #78.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Published detailed plan for phased detection improvements and refinements.

* **Bug Fixes**
  * Improved detection precision from 0.299 to 0.310.
  * Reduced false positive detections in the scanner.
  * Enhanced detection logic to prevent incorrect matches.

[Review Change Stack](https://app.coderabbit.ai/change-stack/recost-dev/extension/pull/102)

### CODE REVIEW

File: docs/superpowers/plans/2026-05-13-a6-a2-a7-detection-fixes.md [hunk 1]
```diff
@@ -0,0 +1,1077 @@
...
+
+NOTE: the exact import paths may need adjusting based on actual API of `scanWorkspace`/`realFilesystemAdapter`. Verify the signatures before running — the scanner CLI uses these so the shape is real. If the API differs, adjust the test wrapper but keep the assertions intact.
+
+- [ ] **Step 2: Wire test into `package.json` test script**
+
+Append to the `test:scanner` script:
+
+```
+&& node dist-test/src/test/a6-object-literal-fps.test.js
+```
```
[the remark this answers is not in the document]
Comment (AndresL230) [reply docs/superpowers/plans/2026-05-13-a6-a2-a7-detection-fixes.md:206]: Skipping. Markdownlint is not enforced in CI here (only `test.yml` and `benchmark.yml` workflows). This is an internal plan doc; other plans in `docs/superpowers/plans/` don't consistently tag code fences either. Will adopt globally if a markdownlint workflow lands.

File: package.json [hunk 2]
```diff
@@ -198,7 +198,7 @@
...
...js && node dist-test/test/ast-cache-detector.test.js && node dist-test/test/ast-batch-detector.test.js && node dist-test/test/ast-concurrency-detector.test.js && node dist-test/test/ast-cross-file-resolver.test.js && node dist-test/intelligence/__tests__/builder.test.js && node dist-test/intelligence/__tests__/clusters.test.js && node dist-test/intelligence/__tests__/compression.test.js && node dist-test/intelligence/__tests__/export.test.js && node dist-test/test/api-client.test.js && node dist-test/test/key-management.test.js && node dist-test/test/ast-parser-loader-fallback.test.js && node dist-test/intelligence/__tests__/cost-utils.test.js && node dist-test/test/intelligence-compression-async.test.js && node dist-test/test/webview-provider-dispatch.test.js && node dist-test/test/extension-activation.test.js && node dist-test/test/source-span.test.js && node dist-test/test/url-template.test.js && node dist-test/test/enclosing-function.test.js && node dist-test/test/endpoint-id.test.js && node dist-test/test/parity.test.js && node dist-test/test/a6-object-literal-fps.test.js && node dist-test/src/test/benchmark-schema.test.js && node dist-test/src/test/benchmark-metrics.test.js",
```
[the remark this answers is not in the document]
Comment (AndresL230) [reply package.json:201]: This is obsolete after commit 6d892ba — the a6 test no longer shells out via `execFileSync` to `dist/cli/scan.js`. It now imports `scanFiles` directly from `core-scanner` (matching the prevailing pattern in `src/test/ast-scanner.test.ts` etc.). So `test:scanner` no longer needs `build:ext` as a prereq. Verified locally: `rm -rf dist && npm test` runs the a6 tests cleanly.

File: src/scanner/core-scanner.ts [hunk 3]
```diff
@@ -193,8 +193,22 @@ export async function scanFiles(
...
+          // and re-emit it as a call at line 1). Fixes A6 / issue #78.
+          const windowEnd = Math.min(lines.length, lineIndex + 6);
+          let astOverlap = false;
+          for (let k = lineIndex; k < windowEnd; k++) {
+            if (astCoveredLines.has(k + 1)) { astOverlap = true; break; }
+          }
+          if (!astOverlap) {
+            const multiLine = lines.slice(lineIndex, windowEnd).join("\n");
+            matches = matchLine(multiLine);
+          }
```
[the remark this answers is not in the document]
Comment (AndresL230) [reply src/scanner/core-scanner.ts:213]: Applied in dcddd72. The narrowed expansion preserves multi-line regex matches that legitimately start before the first AST-covered line. Existing a6 tests still pass (in our fixture the AST-covered line is the call line, so the pre-window slice doesn't contain the chain). Benchmark metrics unchanged — corpus doesn't exercise the specific edge case but the fix is strictly more conservative.
